apps/challenges/aws_utils.py

```diff
@@ -0,0 +1,301 @@
+                                        execution_role_arn=execution_role_arn, container_name=container_name,
+                                        image=image, AWS_DEFAULT_REGION=AWS_DEFAULT_REGION,
+                                        challenge_pk=challenge.pk, log_group=log_group)
+    definition = eval(definition)
+    if challenge.task_def_arn is "":
+        try:
+            response = client.register_task_definition(**definition)
+            task_def_arn = response["taskDefinition"].get("taskDefinitionArn")
+            # To store the task definition arn which will be needed for creating or updating the service later.
+            challenge.task_def_arn = task_def_arn
+            challenge.save()
+            return task_def_arn
+        except Exception as e:
+            logger.info(e)
+            raise
+    error = "Error. Task definition already registered for challenge {}.".format(challenge.pk)
+    raise Exception(error)
+
+
+# Called by the service_manager.
+def create_service_by_challenge_pk(client, challenge, client_token):
+    queue_name = challenge.queue
+    service_name = "{}_service".format(queue_name)
+    if challenge.workers is None:  # Verify if the challenge is new (i.e, service not yet created.).
+        try:
+            task_def_arn = register_task_def_by_challenge_pk(client, queue_name, challenge)
+        except Exception as e:
+            return e.response["Error"]
+
+        definition = service_definition.format(service_name=service_name, task_def_arn=task_def_arn, client_token=client_token)
+        definition = eval(definition)
+        try:
+            response = client.create_service(**definition)
+            # To verify if the service has been created on AWS.
+            if (response["ResponseMetadata"]["HTTPStatusCode"] == 200) and (response["service"]["serviceName"] == service_name):
+                challenge.workers = 1
+                challenge.save()
+                message = {"Success": "Service created succesfully."}
+                return message
+            return "Creation for challenge {}.".format(challenge.pk)
+        except ClientError as e:
+                return e.response["Error"]
+    return "Worker service for challenge {} already exists. Please scale, stop or delete.".format(challenge.pk)
+
+
+# Called by the service_manager.
+def update_service_by_challenge_pk(client, challenge, num_of_tasks):
+    queue_name = challenge.queue
+    service_name = "{}_service".format(queue_name)
+    task_def_arn = challenge.task_def_arn
+
+    definition = update_service_args.format(service_name=service_name, task_def_arn=task_def_arn)
+    definition = eval(definition)
+
+    try:
+        response = client.update_service(**definition)
+        # To verify if the service has been updated on AWS.
+        if (response["ResponseMetadata"]["HTTPStatusCode"] == 200) and (response["service"]["serviceName"] == service_name):
+            challenge.workers = num_of_tasks
```

Comment:
Why are we calling `num_of_workers` as `num_of_tasks` any specific reason?

Reply:
Because in AWS terms, it's the task's desired count we are updating. Just easier to interpret by a third person as to what is being done exactly from the perspective of AWS?
@RishabhJain2018